pkg/testcase/clusterrestore.go

```diff
@@ -0,0 +1,260 @@
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
+
+	remoteDir := fmt.Sprintf("/etc/rancher/%s/", cluster.Config.Product)
+	user := os.Getenv("aws_user")
+	cmd := fmt.Sprintf("sudo mkdir -p %s && sudo chown %s %s ", remoteDir, user, remoteDir)
+
+	_, mkdirCmdErr := shared.RunCommandOnNode(cmd, newClusterIP)
+	Expect(mkdirCmdErr).NotTo(HaveOccurred())
+
+	scpErr := shared.RunScp(cluster, newClusterIP, []string{tempFile.Name()}, []string{remoteDir + "config.yaml"})
+	Expect(scpErr).NotTo(HaveOccurred())
+}
+
+func restoreS3Snapshot(
+	cluster *shared.Cluster,
+	onDemandPath,
+	token string,
+	newClusterIP string,
+	flags *customflag.FlagConfig,
+) {
+	var (
+		restoreCmdRes string
+		restoreCmdErr error
+	)
+
+	productLocationCmd, findErr := shared.FindPath(cluster.Config.Product, newClusterIP)
+	Expect(findErr).NotTo(HaveOccurred())
+
+	restoreCmd := fmt.Sprintf("sudo %s server --cluster-reset --etcd-s3 --cluster-reset-restore-path=%s"+
+		" --etcd-s3-bucket=%s --etcd-s3-folder=%s --etcd-s3-region=%s --etcd-s3-access-key=%s"+
+		" --etcd-s3-secret-key=%s --token=%s",
+		productLocationCmd,
+		onDemandPath,
+		flags.S3Flags.Bucket,
+		flags.S3Flags.Folder,
+		cluster.Aws.Region,
+		cluster.Aws.AccessKeyID,
+		cluster.Aws.SecretAccessKey,
+		token)
+
+	switch cluster.Config.Product {
+	case "k3s":
+		restoreCmdRes, restoreCmdErr = shared.RunCommandOnNode(restoreCmd, newClusterIP)
+		Expect(restoreCmdErr).NotTo(HaveOccurred())
+		Expect(restoreCmdRes).To(ContainSubstring("Managed etcd cluster"))
+		Expect(restoreCmdRes).To(ContainSubstring("has been reset"))
+	case "rke2":
+		_, restoreCmdErr = shared.RunCommandOnNode(restoreCmd, newClusterIP)
+		Expect(restoreCmdErr).To(HaveOccurred())
+		Expect(restoreCmdErr.Error()).To(Not(BeNil()))
+		Expect(restoreCmdErr.Error()).To(ContainSubstring("Managed etcd cluster"))
+		Expect(restoreCmdErr.Error()).To(ContainSubstring("has been reset"))
+	default:
+		Expect(fmt.Errorf("product not supported: %s", cluster.Config.Product)).NotTo(HaveOccurred())
+	}
+}
+
+func enableAndStartService(cluster *shared.Cluster, newClusterIP string) {
+	_, enableServiceCmdErr := shared.ManageService(cluster.Config.Product, "enable", "server",
+		[]string{newClusterIP})
+	Expect(enableServiceCmdErr).NotTo(HaveOccurred())
+
+	_, startServiceCmdErr := shared.ManageService(cluster.Config.Product, "start", "server",
+		[]string{newClusterIP})
+	Expect(startServiceCmdErr).NotTo(HaveOccurred())
+
+	shared.LogLevel("info", "Starting service, waiting for service to complete background processes.")
+
+	status, statusServiceCmdErr := shared.ManageService(cluster.Config.Product, "status", "server",
+		[]string{newClusterIP})
+	Expect(statusServiceCmdErr).NotTo(HaveOccurred())
+	Expect(status).To(ContainSubstring("active "))
+
+	shared.LogLevel("info", "%s service successfully enabled", cluster.Config.Product)
+}
+
+func postValidationRestore(cluster *shared.Cluster, k8sClient *k8s.Client, newServerIP string) {
```

Comment:
nit: `validateAfterS3Restore` or `validatePostS3Restore`

Reply:
lets have geneeric not mentioning s3 is better 
because we are not doing anything with s3 specific related